lib/spack/spack/ci.py

```diff
@@ -678,17 +678,18 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
                 for s in affected_specs:
                     tty.debug('  {0}'.format(s.name))
 
-    generate_job_name = os.environ.get('CI_JOB_NAME', None)
-    parent_pipeline_id = os.environ.get('CI_PIPELINE_ID', None)
+    # Downstream jobs will "need" (depend on, for both scheduling and
+    # artifacts, which include spack.lock file) this pipeline generation
+    # job by both name and pipeline id.  If those environment variables
+    # do not exist, then maybe this is just running in a shell, in which
+    # case, there is no expectation gitlab will ever run the generated
+    # pipeline and those environment variables do not matter.
+    generate_job_name = os.environ.get('CI_JOB_NAME', 'job-does-not-exist')
+    parent_pipeline_id = os.environ.get('CI_PIPELINE_ID', 'pipeline-does-not-exist')
```

Comment:
This function is ~700 lines. Is it feasible, in the short time frame we have before v0.18, to break it in smaller logical chunks?

Reply:
I think anything I try to do this week is going to be rushed and may just have to be revisited again shortly anyway.  If you're ok waiting until after the pending deadlines, there's a lot of refactoring I'd like to tackle in here.

Reply:
I'm ok with holding off on the refactor here, but we really do need to clean this stuff up soon after 0.18, as there's getting to be a lot of tech debt.

I think the interesting question here is how best to separate the stuff we need for production pipelines and for pipelines created by end users.  It seems like the specifics of prod are creeping in here, and I think it makes it harder to generify the code.